File: Problem-2_3Sum.py
# APPROACH 1: Optimal Solution
# Time Complexity : O(n^2), n: length of nums
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. Sort the array - O(n lg n)
# 2. For each element of the array - first number, ensure it's not same as previous one.
# 3. Do Two sum like approach for second and third number.
# 4. If sum is 0, update second and third num such that they are not equal to previous and after elements respectively

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        if not nums or len(nums) < 3:
            return []
        
        result = []
        nums.sort()
        
        for first_ind in range(len(nums) - 2):
            if first_ind > 0 and nums[first_ind - 1] == nums[first_ind]:
                continue
                
            second_ind, third_ind =  first_ind + 1, len(nums) - 1
            
            while second_ind < third_ind:                
                curr_sum = nums[first_ind] + nums[second_ind] + nums[third_ind]
                
                if curr_sum == 0:
                    result.append([nums[first_ind], nums[second_ind], nums[third_ind]])
                    
                    second_ind += 1
                    while second_ind < third_ind and nums[second_ind - 1] == nums[second_ind]:
                        second_ind += 1
                        
                    third_ind -= 1
                    while second_ind < third_ind and nums[third_ind + 1] == nums[third_ind]:
                        third_ind -= 1
                    
                elif curr_sum < 0:
                    second_ind += 1
                    
                elif curr_sum > 0:
                    third_ind -= 1
                    
        return result
                        
                        
        

File: Problem2_3Sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// Time Complexity : O(n^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Sort the array to perform the two pointer approach
// Take ith index element as first number and perform two pointer approach to make the sum 0 on the remaining array
// If i==i+1 skip to the next index, if sum>0 reduce the right pointer else increase the left pointer
// If sum==0, increase l and reduce r - if l or r indeces are having the same elements skip until we get a different element

public class Problem2_3Sum {
    public List<List<Integer>> threeSum(int[] nums) {
        int n=nums.length;
        List<List<Integer>> res=new ArrayList<>();
        Arrays.sort(nums);
        for(int i=0; i<n-2; i++){
            if(i!=0 && nums[i]==nums[i-1]) continue;
            if(nums[i]>0) break;
            int l=i+1, r=n-1;
            while(l<r){
                int sum=nums[i]+nums[l]+nums[r];
                if(sum==0){
                    res.add(Arrays.asList(nums[i], nums[l], nums[r]));
                    l++;
                    r--;
                    while(l<r && nums[l]==nums[l-1]) l++;
                    while(l<r && nums[r]==nums[r+1]) r--;
                }else if(sum>0){
                    r--;
                }else{
                    l++;
                }
            }
        }
        return res;
    }
}
